File: arrays/01_sorted_squared_array/01_sorted_squared_array.js
const testCases = [
  [],
  [0, 5],
  [1, 3, 5],
  [-4, -1, 0, 3, 10],
  [-7, -3, 2, 3, 11],
];

(function () {
  testCases.forEach((testcase, i) => {
    console.log(`testcase ${i}:`);
    console.log(`Sorted Squared Array: ${SortedSquaredArray(testcase)}`);
    console.log(`Two Pointers: ${SquareAndSortOptimized(testcase)}`);
  });
})();

// Approach 1: Sort and Square - T: O(n log n) | S:(1)
function SortedSquaredArray(array) {
  return array
    .map((num) => {
      return num * num;
    })
    .sort((a, b) => {
      return a - b;
    });
}

// Approach2: Two Pointers - T: O(n)
function SquareAndSortOptimized(array) {
  const result = new Array(array.length).fill(0);
  let left = 0;
  let right = array.length - 1;

  for (let i = array.length - 1; i >= 0; i--) {
    let leftSquare = array[left] * array[left];
    let rightSquare = array[right] * array[right];

    if (leftSquare > rightSquare) {
      result[i] = leftSquare;
      left++;
    } else {
      result[i] = rightSquare;
      right--;
    }
  }

  return result;
}

File: arrays/01_sorted_squared_array/go/01_sorted_squared_array.go
package main

import (
	"fmt"
	"sort"
)

func main() {
	testCases := [][]int{
		{},
		{0, 5},
		{1, 3, 5},
		{-4, -1, 0, 3, 10},
		{-7, -3, 2, 3, 11},
	}

	for i, e := range testCases {
		fmt.Printf("TestCase %v:\n", i)
		fmt.Printf("Sorted Squared Array: %v\n", SortedSquaredArray(e))
		fmt.Printf("Tow Pointer: %v\n", SquareAndSortOptimized(e))
	}

}

// Approach 1:
// Sort and Square - T: O(n log n) | S: O(n)
func SortedSquaredArray(array []int) []int {

	result := make([]int, len(array))

	for i, num := range array {
		result[i] = num * num
	}

	sort.Ints(result)
	return result
}

// Using multiplication instead of math.Pow
/*
func square(input int) int {
	return int(math.Pow(float64(input), 2))
}
*/

// Approach 2:
// Two Pointers - T: O(n) | S: O(n)
func SquareAndSortOptimized(array []int) []int {

	result := make([]int, len(array))
	left, right := 0, len(array)-1

	for i := len(array) - 1; i >= 0; i-- {
		leftSquare := array[left] * array[left]
		rightSquare := array[right] * array[right]

		if leftSquare > rightSquare {
			result[i] = leftSquare
			left++
		} else {
			result[i] = rightSquare
			right--
		}
	}

	return result
}
